{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Dotplot.py\n"
     ]
    }
   ],
   "source": [
    "\n",
    "%%writefile Dotplot.py \n",
    "import argparse\n",
    "import multiprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "\n",
    "def merge_sequences_from_fasta(file_path):\n",
    "    sequences = [str(record.seq) for record in SeqIO.parse(file_path, \"fasta\")]\n",
    "    return \"\".join(sequences)\n",
    "\n",
    "def generate_dotplot_seq(seq1, seq2):\n",
    "    matrix = (seq1[:, None] == seq2).astype(np.int8)\n",
    "\n",
    "    plt.imshow(matrix, cmap='Greys', aspect='auto', extent=[0, len(seq2), len(seq1), 0], origin='upper')\n",
    "    plt.xlabel('Sequence 2')\n",
    "    plt.ylabel('Sequence 1')\n",
    "    plt.show()\n",
    "\n",
    "def generate_dotplot_multiprocessing(seq1, seq2, num_processes):\n",
    "    len_seq1 = len(seq1)\n",
    "    len_seq2 = len(seq2)\n",
    "    dotplot_matrix = np.zeros((len_seq1, len_seq2), dtype=bool)\n",
    "\n",
    "    def worker(start, end):\n",
    "        for i in range(start, end):\n",
    "            for j in range(len_seq2):\n",
    "                if seq1[i] == seq2[j]:\n",
    "                    dotplot_matrix[i, j] = 1\n",
    "\n",
    "    processes = []\n",
    "    block_size = len_seq1 // num_processes\n",
    "\n",
    "    for i in range(num_processes):\n",
    "        start = i * block_size\n",
    "        end = start + block_size if i < num_processes - 1 else len_seq1\n",
    "        p = multiprocessing.Process(target=worker, args=(start, end))\n",
    "        processes.append(p)\n",
    "        p.start()\n",
    "\n",
    "    for p in processes:\n",
    "        p.join()\n",
    "\n",
    "    generate_dotplot_seq(seq1, seq2)\n",
    "\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description=\"Dotplot Generator\")\n",
    "    parser.add_argument(\"file1\", help=\"Path to the first FASTA file\")\n",
    "    parser.add_argument(\"file2\", help=\"Path to the second FASTA file\")\n",
    "    parser.add_argument(\"--multiprocessing\", action=\"store_true\", help=\"Use multiprocessing\")\n",
    "\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    seq1 = np.fromiter(merge_sequences_from_fasta(args.file1), dtype=np.dtype('U1'))\n",
    "    seq2 = np.fromiter(merge_sequences_from_fasta(args.file2), dtype=np.dtype('U1'))\n",
    "\n",
    "    if args.multiprocessing:\n",
    "        num_processes = multiprocessing.cpu_count()\n",
    "        generate_dotplot_multiprocessing(seq1, seq2, num_processes)\n",
    "    else:\n",
    "        generate_dotplot_seq(seq1, seq2)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"Dotplot.py\", line 63, in <module>\n",
      "    main()\n",
      "  File \"Dotplot.py\", line 58, in main\n",
      "    generate_dotplot_multiprocessing(seq1, seq2, num_processes)\n",
      "  File \"Dotplot.py\", line 38, in generate_dotplot_multiprocessing\n",
      "    p.start()\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\\lib\\multiprocessing\\process.py\", line 121, in start\n",
      "    self._popen = self._Popen(self)\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\\lib\\multiprocessing\\context.py\", line 224, in _Popen\n",
      "    return _default_context.get_context().Process._Popen(process_obj)\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\\lib\\multiprocessing\\context.py\", line 327, in _Popen\n",
      "    return Popen(process_obj)\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\\lib\\multiprocessing\\popen_spawn_win32.py\", line 93, in __init__\n",
      "    reduction.dump(process_obj, to_child)\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\\lib\\multiprocessing\\reduction.py\", line 60, in dump\n",
      "    ForkingPickler(file, protocol).dump(obj)\n",
      "AttributeError: Can't pickle local object 'generate_dotplot_multiprocessing.<locals>.worker'\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\\lib\\multiprocessing\\spawn.py\", line 107, in spawn_main\n",
      "    new_handle = reduction.duplicate(pipe_handle,\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\\lib\\multiprocessing\\reduction.py\", line 79, in duplicate\n",
      "    return _winapi.DuplicateHandle(\n",
      "PermissionError: [WinError 5] Acceso denegado\n"
     ]
    }
   ],
   "source": [
    "!python Dotplot.py Data\\Chimpanzee.fa Data\\Homosapiens.fa --multiprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import multiprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def read_fasta_file(fasta_file):\n",
    "    sequences = []\n",
    "    with open(fasta_file, \"r\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\">\"):\n",
    "                continue\n",
    "            sequences.append(line.strip())\n",
    "    return sequences\n",
    "\n",
    "def worker(sequence_pair, dotplot_matrix):\n",
    "    sequence1, sequence2 = sequence_pair\n",
    "    result = np.fromiter((x == sequence2 for x in sequence1), dtype=np.int8)\n",
    "    dotplot_matrix += result\n",
    "\n",
    "def make_dotplot(species1_sequences, species2_sequences):\n",
    "    length1 = len(species1_sequences[0])\n",
    "    length2 = len(species2_sequences[0])\n",
    "    manager = multiprocessing.Manager()\n",
    "    dotplot_matrix = manager.list(np.zeros((length1, length2)))\n",
    "    sequence_pairs = [(sequence1, sequence2) for sequence1, sequence2 in zip(species1_sequences, species2_sequences)]\n",
    "    \n",
    "    with multiprocessing.Pool() as pool:\n",
    "        pool.starmap(worker, zip(sequence_pairs, [dotplot_matrix] * len(sequence_pairs)))\n",
    "    \n",
    "    dotplot_matrix = np.array(dotplot_matrix).reshape(length1, length2)\n",
    "    \n",
    "    return dotplot_matrix\n",
    "\n",
    "def main(file1, file2, output_file):\n",
    "    species1_sequences = read_fasta_file(file1)\n",
    "    species2_sequences = read_fasta_file(file2)\n",
    "    dotplot_matrix = make_dotplot(species1_sequences, species2_sequences)\n",
    "    np.savetxt(output_file, dotplot_matrix, fmt=\"%d\")\n",
    "\n",
    "# Rutas de los archivos FASTA\n",
    "file1 = \"Data/Chimpanzee.fa\"\n",
    "file2 = \"Data/Homosapiens.fa\"\n",
    "\n",
    "# Archivo de salida para el dotplot\n",
    "output_file = \"Data/dotplot.txt\"\n",
    "\n",
    "# Llamada a la funci√≥n principal\n",
    "main(file1, file2, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import multiprocessing\n",
    "from Bio import SeqIO\n",
    "\n",
    "def read_fasta_file(fasta_file):\n",
    "    # Lee un archivo FASTA y devuelve una lista de secuencias\n",
    "    sequences = []\n",
    "    with open(fasta_file, \"r\") as f:\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "            sequences.append(str(record.seq))\n",
    "    \n",
    "    return sequences\n",
    "\n",
    "def worker(sequence_pair):\n",
    "    sequence1, sequence2 = sequence_pair\n",
    "    length1 = len(sequence1)\n",
    "    length2 = len(sequence2)\n",
    "    dotplot_matrix = np.zeros((length1, length2), dtype=np.int8)\n",
    "    \n",
    "    for i in range(length1):\n",
    "        for j in range(length2):\n",
    "            if sequence1[i] == sequence2[j]:\n",
    "                dotplot_matrix[i, j] = 1\n",
    "    \n",
    "    return dotplot_matrix\n",
    "\n",
    "def make_dotplot_parallel(file1, file2):\n",
    "    sequences1 = read_fasta_file(file1)\n",
    "    sequences2 = read_fasta_file(file2)\n",
    "    sequence_pairs = [(sequence1, sequence2) for sequence1 in sequences1 for sequence2 in sequences2]\n",
    "\n",
    "    with multiprocessing.Pool() as pool:\n",
    "        results = pool.map(worker, sequence_pairs)\n",
    "\n",
    "    dotplot_matrix = sum(results)\n",
    "    \n",
    "    return dotplot_matrix\n",
    "\n",
    "# Ejemplo de uso\n",
    "fasta_file1 = \"Data/Chimpanzee2.fa\"  # Ruta del primer archivo FASTA\n",
    "fasta_file2 = \"Data/Homosapiens.fa\"  # Ruta del segundo archivo FASTA\n",
    "\n",
    "dotplot_matrix = make_dotplot_parallel(fasta_file1, fasta_file2)\n",
    "print(dotplot_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: Could not find a version that satisfies the requirement tqm (from versions: none)\n",
      "ERROR: No matching distribution found for tqm\n",
      "\n",
      "[notice] A new release of pip available: 22.3.1 -> 23.1.2\n",
      "[notice] To update, run: C:\\Users\\Heosve\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\python.exe -m pip install --upgrade pip\n"
     ]
    }
   ],
   "source": [
    "%pip install tqm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N√∫mero de n√∫cleos l√≥gicos disponibles: 8\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCanceled future for execute_request message before replies were done"
     ]
    }
   ],
   "source": [
    "import multiprocessing\n",
    "import psutil\n",
    "from Bio import SeqIO\n",
    "import time\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Obtener el n√∫mero de n√∫cleos l√≥gicos disponibles\n",
    "num_cores = multiprocessing.cpu_count()\n",
    "\n",
    "print(\"N√∫mero de n√∫cleos l√≥gicos disponibles:\", num_cores)\n",
    "\n",
    "begin = time.time()\n",
    "def divide_fasta_string(fasta_string, num_chunks):\n",
    "    records_per_chunk = (len(fasta_string) + num_chunks - 1) // num_chunks\n",
    "    chunks = [fasta_string[i:i+records_per_chunk] for i in range(0, len(fasta_string), records_per_chunk)]\n",
    "    return chunks\n",
    "\n",
    "def worker(chunk1, chunk2):\n",
    "    matrix = np.zeros((len(chunk1), len(chunk2)), dtype=np.int8)\n",
    "    for i in range(len(chunk1)):\n",
    "        for j in range(len(chunk2)):\n",
    "            if chunk1[i] == chunk2[j]:\n",
    "                matrix[i, j] = 1\n",
    "    return matrix\n",
    "\n",
    "def fasta_file_to_string(file_path):\n",
    "    with open(file_path, \"r\") as file:\n",
    "        fasta_string = \"\".join(line.strip() for line in file if not line.startswith(\">\"))\n",
    "    return fasta_string\n",
    "\n",
    "def save_string_to_file(string, file_path):\n",
    "    with open(file_path, \"w\") as file:\n",
    "        file.write(string)\n",
    "        \n",
    "        \n",
    "def generate_dotplot_parallel(fasta_string1, fasta_string2):\n",
    "    max_memory = psutil.virtual_memory().available\n",
    "    chunk_size = 100 * 1024 * 1024  # 100 MB por chunk\n",
    "    num_chunks = max_memory // chunk_size\n",
    "    chunks1 = divide_fasta_string(fasta_string1, num_chunks)\n",
    "    chunks2 = divide_fasta_string(fasta_string2, num_chunks)\n",
    "\n",
    "    with multiprocessing.Pool() as pool:\n",
    "        results = []\n",
    "        for chunk1 in chunks1:\n",
    "            for chunk2 in chunks2:\n",
    "                result = pool.apply_async(worker, (chunk1, chunk2))\n",
    "                results.append(result.get())\n",
    "\n",
    "    dotplot_matrix = np.concatenate(results)\n",
    "    return dotplot_matrix\n",
    "\n",
    "\n",
    "# Rutas de los archivos FASTA\n",
    "fasta_file1 = \"Data/E_coli.fna\"\n",
    "fasta_file2 = \"Data/Salmonella.fna\"\n",
    "\n",
    "# Convertir los archivos FASTA a strings\n",
    "fasta_string1 = fasta_file_to_string(fasta_file1)\n",
    "fasta_string2 = fasta_file_to_string(fasta_file2)\n",
    "\n",
    "# Guardar los strings en archivos\n",
    "output_file1 = \"fasta1.txt\"\n",
    "output_file2 = \"fasta2.txt\"\n",
    "save_string_to_file(fasta_string1, output_file1)\n",
    "save_string_to_file(fasta_string2, output_file2)\n",
    "\n",
    "# Generar el dotplot en paralelo\n",
    "begin = time.time()\n",
    "dotplot_matrix = generate_dotplot_parallel(fasta_string1, fasta_string2)\n",
    "execution_time = time.time() - begin\n",
    "\n",
    "# Graficar el dotplot\n",
    "plt.imshow(dotplot_matrix, cmap='Greys', aspect='auto', origin='upper')\n",
    "plt.xlabel('Sequence 2')\n",
    "plt.ylabel('Sequence 1')\n",
    "plt.show()\n",
    "\n",
    "print(f\"\\n El c√≥digo se ejecut√≥ en: {time.time() - begin} segundos\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import multiprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "\n",
    "def calculate_dotplot(sequence1, sequence2):\n",
    "    length1 = len(sequence1)\n",
    "    length2 = len(sequence2)\n",
    "    dotplot_matrix = np.zeros((length1, length2), dtype=np.int8)\n",
    "    \n",
    "    # Crear diccionario para indexar elementos de sequence2\n",
    "    sequence2_dict = {element: i for i, element in enumerate(sequence2)}\n",
    "    \n",
    "    # Recorrer sequence1 y verificar presencia en sequence2_dict\n",
    "    for i in range(length1):\n",
    "        for j in range(length2):\n",
    "            if sequence1[i] == sequence2_dict.get(sequence1[i]):\n",
    "                dotplot_matrix[i, j] = 1\n",
    "    \n",
    "    return dotplot_matrix\n",
    "\n",
    "def make_dotplot_parallel(species1_fragments, species2_fragments):\n",
    "    sequence_pairs = [(sequence1, sequence2) for sequence1 in species1_fragments for sequence2 in species2_fragments]\n",
    "    \n",
    "    with multiprocessing.Pool() as pool:\n",
    "        dotplot_matrices = pool.map(calculate_dotplot, sequence_pairs)\n",
    "    \n",
    "    dotplot_matrix = np.sum(dotplot_matrices, axis=0)\n",
    "    \n",
    "    return dotplot_matrix\n",
    "\n",
    "def load_sequences(file_path):\n",
    "    sequences = []\n",
    "    with open(file_path, \"r\") as file:\n",
    "        for record in SeqIO.parse(file, \"fasta\"):\n",
    "            sequence = str(record.seq)\n",
    "            sequences.append(sequence)\n",
    "    return sequences\n",
    "\n",
    "def split_sequences(sequences, fragment_size):\n",
    "    fragmented_sequences = []\n",
    "    for sequence in sequences:\n",
    "        fragmented_sequence = [sequence[i:i+fragment_size] for i in range(0, len(sequence), fragment_size)]\n",
    "        fragmented_sequences.extend(fragmented_sequence)\n",
    "    return fragmented_sequences\n",
    "\n",
    "def main(file1, file2):\n",
    "    # Cargar las secuencias de los archivos FASTA\n",
    "    species1_sequences = load_sequences(file1)\n",
    "    species2_sequences = load_sequences(file2)\n",
    "\n",
    "    # Dividir las secuencias en fragmentos m√°s peque√±os\n",
    "    fragment_size = 1000\n",
    "    species1_fragments = split_sequences(species1_sequences, fragment_size)\n",
    "    species2_fragments = split_sequences(species2_sequences, fragment_size)\n",
    "\n",
    "    # Calcular el dotplot en paralelo\n",
    "    dotplot_matrix = make_dotplot_parallel(species1_fragments, species2_fragments)\n",
    "\n",
    "    # Continuar con la visualizaci√≥n o el an√°lisis del dotplot...\n",
    "\n",
    "# Rutas de los archivos FASTA\n",
    "file1 = \"Data/Chimpanzee.fa\"\n",
    "file2 = \"Data/Homosapiens.fa\"\n",
    "\n",
    "# Llamada a la funci√≥n principal\n",
    "main(file1, file2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from numba import njit\n",
    "from multiprocessing import Pool, cpu_count\n",
    "\n",
    "def merge_sequences_from_fasta(file_path):\n",
    "    sequences = []  # List to store all sequences\n",
    "    for record in SeqIO.parse(file_path, \"fasta\"):\n",
    "        # `record.seq` gives the sequence\n",
    "        sequences.append(str(record.seq))\n",
    "    return \"\".join(sequences)\n",
    "\n",
    "@njit\n",
    "def generate_dotplot_seq(seq1, seq2, start1, end1, start2, end2, matrix):\n",
    "    for i in range(start1, end1):\n",
    "        for j in range(start2, end2):\n",
    "            if seq1[i] == seq2[j]:\n",
    "                matrix[i - start1, j - start2] = 1\n",
    "\n",
    "def process_section(args):\n",
    "    seq1, seq2, start1, end1, start2, end2 = args\n",
    "    matrix = np.zeros((end1 - start1, end2 - start2), dtype=np.uint8)\n",
    "    generate_dotplot_seq(seq1, seq2, start1, end1, start2, end2, matrix)\n",
    "    return matrix\n",
    "\n",
    "# Rutas de archivo\n",
    "file_path_1 = \"Data/E_coli.fna\"\n",
    "file_path_2 = \"Data/Salmonella.fna\"\n",
    "\n",
    "# Cargar secuencias\n",
    "seq1 = merge_sequences_from_fasta(file_path_1)\n",
    "seq2 = merge_sequences_from_fasta(file_path_2)\n",
    "\n",
    "# Definir el n√∫mero de secciones en las que se dividir√°n las secuencias\n",
    "num_sections = 1000\n",
    "# Calcular el tama√±o de cada secci√≥n\n",
    "section_size1 = len(seq1) // num_sections\n",
    "section_size2 = len(seq2) // num_sections\n",
    "\n",
    "# Crear argumentos para cada secci√≥n\n",
    "args_list = []\n",
    "for i in range(num_sections):\n",
    "    start1 = i * section_size1\n",
    "    end1 = (i + 1) * section_size1\n",
    "    start2 = i * section_size2\n",
    "    end2 = (i + 1) * section_size2\n",
    "    args_list.append((seq1, seq2, start1, end1, start2, end2))\n",
    "\n",
    "# Procesar cada secci√≥n de las secuencias en paralelo utilizando 8 procesos\n",
    "with Pool(processes=4) as pool:\n",
    "    matrices = pool.map(process_section, args_list)\n",
    "\n",
    "# Mostrar los dotplots de cada secci√≥n\n",
    "for i, matrix in enumerate(matrices):\n",
    "    plt.imshow(matrix, cmap='Greys', aspect='auto')\n",
    "    plt.xlabel('Sequence 2')\n",
    "    plt.ylabel('Sequence 1')\n",
    "    plt.title('Part {}'.format(i + 1))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "Out of memory",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[6], line 38\u001b[0m\n\u001b[0;32m     35\u001b[0m seq2 \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mnext\u001b[39m(SeqIO\u001b[39m.\u001b[39mparse(file_path_2, \u001b[39m\"\u001b[39m\u001b[39mfasta\u001b[39m\u001b[39m\"\u001b[39m))\u001b[39m.\u001b[39mseq)\n\u001b[0;32m     37\u001b[0m \u001b[39m# Comparar secuencias y generar el dotplot\u001b[39;00m\n\u001b[1;32m---> 38\u001b[0m aligned_seq1, aligned_seq2 \u001b[39m=\u001b[39m compare_sequences(seq1, seq2)\n\u001b[0;32m     39\u001b[0m generate_dotplot(aligned_seq1, aligned_seq2)\n",
      "Cell \u001b[1;32mIn[6], line 7\u001b[0m, in \u001b[0;36mcompare_sequences\u001b[1;34m(seq1, seq2)\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcompare_sequences\u001b[39m(seq1, seq2):\n\u001b[1;32m----> 7\u001b[0m     alignments \u001b[39m=\u001b[39m pairwise2\u001b[39m.\u001b[39;49malign\u001b[39m.\u001b[39;49mglobalxx(seq1, seq2)\n\u001b[0;32m      8\u001b[0m     best_alignment \u001b[39m=\u001b[39m alignments[\u001b[39m0\u001b[39m]\n\u001b[0;32m     10\u001b[0m     aligned_seq1 \u001b[39m=\u001b[39m best_alignment\u001b[39m.\u001b[39mseqA\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\Bio\\pairwise2.py:502\u001b[0m, in \u001b[0;36malign.alignment_function.__call__\u001b[1;34m(self, *args, **keywds)\u001b[0m\n\u001b[0;32m    500\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Call the alignment instance already created.\"\"\"\u001b[39;00m\n\u001b[0;32m    501\u001b[0m keywds \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdecode(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkeywds)\n\u001b[1;32m--> 502\u001b[0m \u001b[39mreturn\u001b[39;00m _align(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkeywds)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\Bio\\pairwise2.py:569\u001b[0m, in \u001b[0;36m_align\u001b[1;34m(sequenceA, sequenceB, match_fn, gap_A_fn, gap_B_fn, penalize_extend_when_opening, penalize_end_gaps, align_globally, gap_char, force_generic, score_only, one_alignment_only)\u001b[0m\n\u001b[0;32m    567\u001b[0m     open_A, extend_A \u001b[39m=\u001b[39m gap_A_fn\u001b[39m.\u001b[39mopen, gap_A_fn\u001b[39m.\u001b[39mextend\n\u001b[0;32m    568\u001b[0m     open_B, extend_B \u001b[39m=\u001b[39m gap_B_fn\u001b[39m.\u001b[39mopen, gap_B_fn\u001b[39m.\u001b[39mextend\n\u001b[1;32m--> 569\u001b[0m     matrices \u001b[39m=\u001b[39m _make_score_matrix_fast(\n\u001b[0;32m    570\u001b[0m         sequenceA,\n\u001b[0;32m    571\u001b[0m         sequenceB,\n\u001b[0;32m    572\u001b[0m         match_fn,\n\u001b[0;32m    573\u001b[0m         open_A,\n\u001b[0;32m    574\u001b[0m         extend_A,\n\u001b[0;32m    575\u001b[0m         open_B,\n\u001b[0;32m    576\u001b[0m         extend_B,\n\u001b[0;32m    577\u001b[0m         penalize_extend_when_opening,\n\u001b[0;32m    578\u001b[0m         penalize_end_gaps,\n\u001b[0;32m    579\u001b[0m         align_globally,\n\u001b[0;32m    580\u001b[0m         score_only,\n\u001b[0;32m    581\u001b[0m     )\n\u001b[0;32m    582\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    583\u001b[0m     matrices \u001b[39m=\u001b[39m _make_score_matrix_generic(\n\u001b[0;32m    584\u001b[0m         sequenceA,\n\u001b[0;32m    585\u001b[0m         sequenceB,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    591\u001b[0m         score_only,\n\u001b[0;32m    592\u001b[0m     )\n",
      "\u001b[1;31mMemoryError\u001b[0m: Out of memory"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "\n",
    "def compare_sequences(seq1, seq2):\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "    best_alignment = alignments[0]\n",
    "    \n",
    "    aligned_seq1 = best_alignment.seqA\n",
    "    aligned_seq2 = best_alignment.seqB\n",
    "    \n",
    "    return aligned_seq1, aligned_seq2\n",
    "\n",
    "def generate_dotplot(seq1, seq2):\n",
    "    matrix = np.zeros((len(seq1), len(seq2)), dtype=np.uint8)\n",
    "    \n",
    "    for i in range(len(seq1)):\n",
    "        for j in range(len(seq2)):\n",
    "            if seq1[i] == seq2[j]:\n",
    "                matrix[i, j] = 1\n",
    "    \n",
    "    plt.imshow(matrix, cmap='Greys', aspect='auto')\n",
    "    plt.xlabel('Sequence 2')\n",
    "    plt.ylabel('Sequence 1')\n",
    "    plt.title('Dotplot')\n",
    "    plt.show()\n",
    "\n",
    "# Rutas de archivo\n",
    "file_path_1 = \"Data/E_coli.fna\"\n",
    "file_path_2 = \"Data/Salmonella.fna\"\n",
    "\n",
    "# Leer la primera secuencia desde los archivos FASTA\n",
    "seq1 = str(next(SeqIO.parse(file_path_1, \"fasta\")).seq)\n",
    "seq2 = str(next(SeqIO.parse(file_path_2, \"fasta\")).seq)\n",
    "\n",
    "# Comparar secuencias y generar el dotplot\n",
    "aligned_seq1, aligned_seq2 = compare_sequences(seq1, seq2)\n",
    "generate_dotplot(aligned_seq1, aligned_seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from multiprocessing import Pool\n",
    "\n",
    "def merge_sequences_from_fasta(file_path):\n",
    "    sequences = []  # List to store all sequences\n",
    "    for record in SeqIO.parse(file_path, \"fasta\"):\n",
    "        # `record.seq` gives the sequence\n",
    "        sequences.append(str(record.seq))\n",
    "    return \"\".join(sequences)\n",
    "\n",
    "def generate_dotplot_seq(seq1, seq2, start1, end1, start2, end2, matrix):\n",
    "    for i in range(start1, end1):\n",
    "        for j in range(start2, end2):\n",
    "            if seq1[i] == seq2[j]:\n",
    "                matrix[i - start1, j - start2] = 1\n",
    "                \n",
    "def process_section(section_index):\n",
    "    start1 = section_index * section_size1\n",
    "    end1 = (section_index + 1) * section_size1\n",
    "    start2 = section_index * section_size2\n",
    "    end2 = (section_index + 1) * section_size2\n",
    "    \n",
    "    matrix = np.zeros((end1 - start1, end2 - start2), dtype=np.uint8)\n",
    "    generate_dotplot_seq(seq1, seq2, start1, end1, start2, end2, matrix)\n",
    "    \n",
    "    plt.imshow(matrix, cmap='Greys', aspect='auto')\n",
    "    plt.xlabel('Sequence 2')\n",
    "    plt.ylabel('Sequence 1')\n",
    "    plt.title('Part {}'.format(section_index + 1))\n",
    "    \n",
    "    # Save the image in the output folder with a unique filename\n",
    "    filename = os.path.join(output_folder, 'dotplot_{}.png'.format(section_index + 1))\n",
    "    plt.savefig(filename)\n",
    "    \n",
    "    # Clear the current figure to avoid overlapping plots\n",
    "    plt.clf()\n",
    "    plt.close()\n",
    "\n",
    "# Folder to save the dot plot images\n",
    "output_folder = \"imagesgonorrea\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# Rutas de archivo\n",
    "file_path_1 = \"Data/E_coli.fna\"\n",
    "file_path_2 = \"Data/Salmonella.fna\"\n",
    "\n",
    "# Cargar secuencias\n",
    "seq1 = merge_sequences_from_fasta(file_path_1)\n",
    "seq2 = merge_sequences_from_fasta(file_path_2)\n",
    "\n",
    "# Definir el n√∫mero de secciones en las que se dividir√°n las secuencias\n",
    "num_sections = 1000\n",
    "# Calcular el tama√±o de cada secci√≥n inicial\n",
    "section_size1 = len(seq1) // num_sections\n",
    "section_size2 = len(seq2) // num_sections\n",
    "\n",
    "# Crear una pool de procesos con el n√∫mero de procesos deseados\n",
    "pool = Pool(processes=4)\n",
    "\n",
    "# Procesar cada secci√≥n de las secuencias en paralelo\n",
    "pool.map(process_section, range(num_sections))\n",
    "\n",
    "pool.close()\n",
    "pool.join()\n",
    "\n",
    "print(\"Dot plot images saved in the '{}' folder.\".format(output_folder))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
